### PR TITLE
WIP: 🌱 added tests for loadTLSConfigFromEnv in factory_test.go

### LINKED ISSUE #1574: Unit test coverage improvement in BMO
Improve the unit test coverage by writing new test cases or completely new tests. Check the current coverage with `make unit-cover`.

Here are some suggestions on where to start. Note that some functions are harder to test than others because they may depend on access to network resources or a database. If you come across such function, skip them and aim for the low hanging fruits instead.

- [x] [features.go](https://github.com/metal3-io/baremetal-operator/blob/main/pkg/provisioner/ironic/clients/features.go) has no tests at all and looks quite simple to test. Fixed in https://github.com/metal3-io/baremetal-operator/pull/1579
- [ ] [loadTLSConfigFromEnv](https://github.com/metal3-io/baremetal-operator/blob/main/pkg/provisioner/ironic/factory.go#L169) is untested and should be easy to test. Fixed in https://github.com/metal3-io/baremetal-operator/pull/1589 or https://github.com/metal3-io/baremetal-operator/pull/2013
- [x] [baremetalhost_types.go](https://github.com/metal3-io/baremetal-operator/blob/main/apis/metal3.io/v1alpha1/baremetalhost_types.go) has multiple small functions without coverage that could easily be tested. Fixed in https://github.com/metal3-io/baremetal-operator/pull/1615
- [x] [bmceventsubscription_controller.go](https://github.com/metal3-io/baremetal-operator/blob/main/controllers/metal3.io/bmceventsubscription_controller.go) has no tests at all. Fixed in https://github.com/metal3-io/baremetal-operator/pull/1615
- [ ] [preprovisioningimage_controller.go](https://github.com/metal3-io/baremetal-operator/blob/main/controllers/metal3.io/preprovisioningimage_controller.go) has no tests at all. Fixed in https://github.com/metal3-io/baremetal-operator/pull/2013
- [x] [LoadAuth](https://github.com/metal3-io/baremetal-operator/blob/b5c4120dd294e94cd63fede91321f38623293ec7/pkg/provisioner/ironic/clients/auth.go#L43) is untested. Fixed in https://github.com/metal3-io/baremetal-operator/pull/1774
- [x] [client.go](https://github.com/metal3-io/baremetal-operator/blob/main/pkg/provisioner/ironic/clients/client.go) has no tests at all. Fixed in https://github.com/metal3-io/baremetal-operator/pull/1606

### PR DESCRIPTION
**Pull request by danish9039**
This commit introduces a set of comprehensive tests for the in theloadingTLSconfiguration in"factory_test.go" file of the ironic package. The tests cover various scenarios and edge cases related to loading configuration from environment variables, handling endpoints, and loading TLSconfiguration.

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
part of #1574

### CODE REVIEW

**Comment by tuminoid** [comment]
Please sign-off your commit to fulfill DCO requirement.

/ok-to-test

**Comment by lentzi90** [comment]
Thanks for the PR!
You will need to fix the test failures. Try running `make lint` to see the issues in your environment also.

**Comment by lentzi90** [comment]
Could you update the description so it says "Part of #1574" instead of "Fixes", since the issue has multiple parts?

File: pkg/provisioner/ironic/factory_test.go [hunk 1]
```diff
@@ -181,35 +181,78 @@ func TestLoadConfigFromEnv(t *testing.T) {
 	}
 }
 
-func TestLoadEndpointsFromEnv(t *testing.T) {
```
Comment (lentzi90) [inline pkg/provisioner/ironic/factory_test.go:184]: Do I understand correctly that you removed this test? Why?
I think we should keep it and add a separate test for the TLS config.

**Comment by danish9039** [comment]
> Could you update the description so it says "Part of #1574" instead of "Fixes", since the issue has multiple parts?

sure

File: pkg/provisioner/ironic/factory_test.go [hunk 2]
```diff
@@ -181,35 +181,78 @@ func TestLoadConfigFromEnv(t *testing.T) {
...
+	os.Setenv("IRONIC_CLIENT_PRIVATE_KEY_FILE", "/path/to/client.key")
+	os.Setenv("IRONIC_INSECURE", "false")
+	os.Setenv("IRONIC_SKIP_CLIENT_SAN_VERIFY", "true")
+	defer func() {
+		os.Unsetenv("IRONIC_CACERT_FILE")
+		os.Unsetenv("IRONIC_CLIENT_CERT_FILE")
+		os.Unsetenv("IRONIC_CLIENT_PRIVATE_KEY_FILE")
+		os.Unsetenv("IRONIC_INSECURE")
+		os.Unsetenv("IRONIC_SKIP_CLIENT_SAN_VERIFY")
+	}()
```
Comment (lentzi90) [inline pkg/provisioner/ironic/factory_test.go:201]: If you check the other tests in this file, you will see that they make use of a struct [EnvFixture](https://github.com/metal3-io/baremetal-operator/blob/aef07f5593ae48d589275d88cd43015c9171d7e7/pkg/provisioner/ironic/factory_test.go#L10-L18). There are [helpful functions](https://github.com/metal3-io/baremetal-operator/blob/aef07f5593ae48d589275d88cd43015c9171d7e7/pkg/provisioner/ironic/factory_test.go#L20-L46) for setting and unsetting the environment from this struct. Put all of this together and you get a nice way of looping through test cases described by structs. Could you try to mimic this?

**Comment by tuminoid** [comment]
/retitle WIP: 🌱 added tests for loadTLSConfigFromEnv in factory_test.go

There are open questions from @lentzi90 that require changes to the PR, marking as WIP.

**Comment by Rozzii** [comment]
IMO we should leave the "stale" on this until the questions from @lentzi90 are answered.
Are you still willing to work on this @danish9039  ?
